### PR TITLE
Issue #3 - New Shopping List

### LINKED ISSUE #3: 3. As a user, I want to set up a new shopping list so I can start tracking purchased items.
A shopping list consists of a set of items associated with a user’s token. Creating a new list consists of the following sequence:

* Generate a new, unique token
* Save the token to `localStorage`
* Show the user the list view

The script `getToken` from the `shopping-list-utils` package should be used to generate the token. More information on that package [here](https://www.npmjs.com/package/@the-collab-lab/shopping-list-utils).

`import { getToken, words } from '@the-collab-lab/shopping-list-utils'`

AC:

* `@the-collab-lab/shopping-list-utils` has been added as a dependency to the project.
* For users who do not already have a token or a list, a button or link exists on the home screen that allows them to create a new list
* Clicking the button/link generates a new token and saves it to `localStorage`
* Once the token has been created and saved, the user is redirected to the "list" view
* For users who do have a token already saved in `localStorage`, they should be shown/redirected to the "list" view

Notes:

* Eventually the welcome/home page will look like the following wireframe: https://github.com/the-collab-lab/smart-shopping-list/blob/main/_resources/wireframes/welcome-screen.png
* It's good for your whole team to look at the wireframe now, but be sure not to try to implement all of the visible functionality at once. There will be future stories to add the other features such as joining an existing list.
* For this story, it's sufficient just to have a simple "create list" button. Although you are not required to save the token to Firestore, some teams have opted to also save the token to Firestore when generating a token.

### PR DESCRIPTION
**Pull request by rjwbeard**
## Description

We added the `@the-collab-lab/shopping-list-utils` package as a project dependency and used the `getToken` function to generate user tokens which we then save to `localStorage`. We then wrote a Welcome component that renders a heading and a simple button if a token is not found in `localStorage`. If there is a token present, the user is redirected to their existing shopping list.

We used dynamic routes in the navigation in order to pass the token to other components with `useParams`. We wrote two tests; one for the component rendering and one for the button push correctly modeling the localStorage functionality.

## Related Issue
closes #3 

## Acceptance Criteria
- [x] @the-collab-lab/shopping-list-utils has been added as a dependency to the project.

- [x] For users who do not already have a token or a list, a button or link exists on the home screen that allows them to create a new list

- [x] Once the token has been created and saved, the user is redirected to the "list" view
    
- [x] Clicking the button/link generates a new token and saves it to localStorage

- [x] For users who do have a token already saved in localStorage, they should be shown/redirected to the "list" view

## Type of Changes
|     | Type                       |
| --- | -------------------------- |
|  ✓ | :sparkles: New feature     |
|   ✓ | :100: Add tests            |
|  ✓ | :link: Update dependencies |

### Welcome Component
Screen Shot 2022-04-13 at 12 05 36 PM

## Testing Steps / QA Criteria

- `git pull origin rb-tt-new-shopping-list` to pull down the branch
- `npm install` to update dependencies
- `npm start` to run the branch locally
- `npm test` to run the test files

### CODE REVIEW

**Review by timothy-taylor — COMMENTED** [review]
I wanted to note the purpose of the `parseToken` util function was to fix a bug created by using dynamic routes with `<NavLink/>`. NavLink works by comparing the supplied string with the browser url path. Since the token is three words separated by spaces, and the browser converts spaces to `%20`, we end up comparing `"some random words"` to `"some%20random%20words"` and our boolean check fails and the active Link is not correctly styled. The `parseToken` function simply replaces the spaces with a "-" so our comparison ends up being `"some-random-words" === "some-random-words"` and the built in styling mechanism in `<NavLink>` works again.

File: src/components/Welcome/Welcome.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    window.localStorage.getItem('user-token'),
+  );
+
+  const handleClick = (e) => {
+    e.preventDefault();
+    const parsedToken = parseToken(getToken(words));
+    setToken(parsedToken);
+  };
+
+  React.useEffect(() => {
```
Comment (shirenekboyd) [inline src/components/Welcome/Welcome.js:18]: I haven't seen this type of useEffect before, very cool!

File: src/components/Welcome/Welcome.test.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+   <MemoryRouter>
+    <Welcome />
+   </MemoryRouter>,
+  );
+ 
+  const button = screen.getByRole('button', { name: "Create new list" });
+  userEvent.click(button);
+ 
+  expect(localStorage.setItem).toHaveBeenCalled();
+ });
```
Comment (shirenekboyd) [inline src/components/Welcome/Welcome.test.js:39]: You are amazing with implementing tests!  I need to learn how to do this better.

**Review by EmilyDSarani — COMMENTED** [review]
Love what you guys did here. It has been so long since i've done anything with tokens, so it was really cool to see code like this again.

**Review by An-dy1 — APPROVED** [review]
Well done you two, and good test coverage. Everything is behaving as expected, including when there is already a token in storage. Keep up the good work!

**Pull request merged**